File: frontend/src/jmix/entities/Spaceport.ts
import { Planet } from "./Planet";
import { Moon } from "./Moon";
import { Carrier } from "./Carrier";
export class Spaceport {
  static NAME = "Spaceport";
  id?: string;
  name?: string | null;
  planet?: Planet | null;
  moon?: Moon | null;
  coordinates?: any | null;
  isDefault?: boolean | null;
  carriers?: Carrier[] | null;
}
export type SpaceportViewName =
  | "_base"
  | "_instance_name"
  | "_local"
  | "spaceport-fetch-plan";
export type SpaceportView<V extends SpaceportViewName> = V extends "_base"
  ? Pick<Spaceport, "id" | "name" | "isDefault">
  : V extends "_instance_name"
  ? Pick<Spaceport, "id" | "name">
  : V extends "_local"
  ? Pick<Spaceport, "id" | "name" | "isDefault">
  : V extends "spaceport-fetch-plan"
  ? Pick<
      Spaceport,
      "id" | "name" | "isDefault" | "planet" | "moon" | "carriers"
    >
  : never;

File: src/main/java/com/company/spacetrans/entity/Spaceport.java
package com.company.spacetrans.entity;

import io.jmix.core.entity.annotation.EmbeddedParameters;
import io.jmix.core.entity.annotation.JmixGeneratedValue;
import io.jmix.core.metamodel.annotation.InstanceName;
import io.jmix.core.metamodel.annotation.JmixEntity;

import javax.persistence.*;
import java.util.List;
import java.util.UUID;

@JmixEntity
@Table(name = "SPACEPORT")
@Entity
public class Spaceport {
    @JmixGeneratedValue
    @Column(name = "ID", nullable = false)
    @Id
    private UUID id;

    @InstanceName
    @Column(name = "NAME")
    private String name;

    @JoinColumn(name = "PLANET_ID")
    @ManyToOne(fetch = FetchType.LAZY)
    private Planet planet;

    @JoinColumn(name = "MOON_ID")
    @ManyToOne(fetch = FetchType.LAZY)
    private Moon moon;

    @Embedded
    @EmbeddedParameters(nullAllowed = false)
    @AttributeOverrides({
            @AttributeOverride(name = "latitude", column = @Column(name = "COORDINATES_LATITUDE")),
            @AttributeOverride(name = "longitude", column = @Column(name = "COORDINATES_LONGITUDE"))
    })
    private Coordinates coordinates;

    @Column(name = "IS_DEFAULT")
    private Boolean isDefault;

    @JoinTable(name = "CARRIER_SPACEPORT_LINK",
            joinColumns = @JoinColumn(name = "SPACEPORT_ID"),
            inverseJoinColumns = @JoinColumn(name = "CARRIER_ID"))
    @ManyToMany
    private List<Carrier> carriers;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Boolean getIsDefault() {
        return isDefault;
    }

    public void setIsDefault(Boolean isDefault) {
        this.isDefault = isDefault;
    }

    public Coordinates getCoordinates() {
        return coordinates;
    }

    public void setCoordinates(Coordinates coordinates) {
        this.coordinates = coordinates;
    }

    public List<Carrier> getCarriers() {
        return carriers;
    }

    public void setCarriers(List<Carrier> carriers) {
        this.carriers = carriers;
    }

    public Moon getMoon() {
        return moon;
    }

    public void setMoon(Moon moon) {
        this.moon = moon;
    }

    public Planet getPlanet() {
        return planet;
    }

    public void setPlanet(Planet planet) {
        this.planet = planet;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }
}